core/src/main/java/ai/timefold/solver/core/impl/score/director/ValueRangeState.java

```diff
@@ -0,0 +1,98 @@
+package ai.timefold.solver.core.impl.score.director;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import ai.timefold.solver.core.api.domain.valuerange.CountableValueRange;
+import ai.timefold.solver.core.api.domain.valuerange.ValueRange;
+import ai.timefold.solver.core.api.solver.change.ProblemChange;
+import ai.timefold.solver.core.impl.domain.valuerange.descriptor.ValueRangeDescriptor;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Caches value ranges for the current working solution,
+ * allowing to quickly access.
+ * 
+ * <p>
+ * Outside a {@link ProblemChange}, value ranges are not allowed to change.
+ * Call {@link #reset()} every time the working solution changes through a problem fact,
+ * so that all caches can be invalidated.
+ *
+ * @see ValueRange
+ */
+@NullMarked
+public final class ValueRangeState<Solution_> implements ValueRangeResolver<Solution_> {
```

Comment:
For a single-impl interfaces, I think the interface just creates needless complications. Keep the name of the interface for the class, it is better than the actual name of the class, but remove the interface.